### ISSUE TITLE
Customize "composer install"

### ISSUE THREAD

**Issue by mbrodala** [nanobox-io/nanobox-engine-php#19]
Currently the following command is executed in the [`Installing binaries and runtimes` step](https://github.com/nanobox-io/nanobox-engine-php/blob/35a3cbe/lib/php/composer.sh#L21):

`composer install --no-interaction --prefer-source`

I'd love to have an option to customize this. For development I need this:

`composer install --no-interaction`

For production I need this:

`composer install --no-interaction --no-dev --prefer-dist --optimize-autoloader`

**Comment by mbrodala** [comment]
@tylerflint @notxarb could you have another look at this? Currently starting boxes takes ages since every package is cloned instead of simply downloaded via dist.

**Comment by mbrodala** [comment]
I've just applied [this change to my fork](https://github.com/mbrodala/nanobox-engine-php/commit/92bda3bb386d73040aa69dac3dc95ae240ea1cb7) and box startup is now **a lot** faster.

For now I'll use `engine: mbrodala/nanobox-engine-php` in my `boxfile.yml` but of course, this only a temporary solution.

**Comment by tylerflint** [comment]
@mbrodala This looks good to me. I'm curious why we have the --prefer-dist in the command by default. @notxarb do you know?

**Comment by notxarb** [comment]
From composer docs:
`--prefer-dist:` Reverse of --prefer-source, Composer will install from dist if possible. This can speed up installs substantially on build servers and other use cases where you typically do not run updates of the vendors. It is also a way to circumvent problems with git if you do not have a proper setup.
I think we ran into issues of asking if you wanted to add a key to the known hosts file, or if the ssh key has a passphrase. The prefer-dist avoids those issues.

**Comment by kristoftorfs** [comment]
Any updates on this? How about allowing customization through the boxfile.yml?

**Comment by tylerflint** [comment]
That's a great idea. @notxarb any objections?

**Comment by notxarb** [comment]
I don't have any objections to allowing a custom command for composer

**Comment by tim-bec** [comment]
@notxarb please rethink about that. we use external projects where composer fails without `--no-dev` I know that the issue needs to be fixed in the project itself - but as a workaround it would be great to add custom commands or run composer by hand in the nandbox.

**Comment by sanderson** [comment]
I think this could be a fairly simple change. We could leave the current `composer install` command as a reasonable default, but let users override it with a boxfile.yml config where they just pass custom composer install arguments and flags. This would only customize the `composer install` command that runs during the build. All other composer commands would just need be added as extra_steps.

```yaml
run.config:
  engine: php
  engine.config:
    composer_install: --no-dev --prefer-source --optimize-autoloader
  extra_steps:
    - composer dump-autoload
```

**Comment by mbrodala** [comment]
@sanderson I'm not sure this is enough, at least I don't see how I could set two different commands for local and build.

And `composer dump-autoload` is not necessary, `composer install` does this automatically. ;-)

**Comment by danhunsaker** [comment]
Local is what we call "build", whereas production is what we call "compile". For the compile phase, you simply add:

```yaml
deploy.config:
  extra_steps:
    - composer install --no-interaction --no-dev --prefer-dist --optimize-autoloader
```
and Composer will clean up your dependencies at that point.

**Comment by sanderson** [comment]
Thanks @mbrodala, I was just dropping random flags in as an example. @danhunsaker, is the `deploy.config > extra_steps > composer ...` included by the engine? If not, I wonder if it should be.

**Comment by danhunsaker** [comment]
I don't believe it is, no. Not all projects care about dev versus prod dependencies, but re-running the install command during compile, with `--no-dev` added, is very much a sane default.

**Comment by kristoftorfs** [comment]
I also think it would be good practice to put in there, people who care about dev/prod dependencies no longer have two include extra's in their `boxyfile.yml` while the people who don't care wel... they don't care.

Another way to make it easier would be to [allow importing of snippets](https://github.com/nanobox-io/nanobox/issues/589) into our `boxfile.yml`. 😄

**Comment by mbrodala** [comment]
@danhunsaker Good idea, the overhead should be minimal since Composer only needs to remove packages which where installed on build already.

**Comment by asenor** [comment]
Hi all! I would like to know the status of this feature. The implementation on feature/custom-composer looks pretty good!

Also I would like to ask if it will be possible to add [hirak/prestissimo](https://github.com/hirak/prestissimo) globally to speed up composer install.

Thanks for your work!

**Comment by danhunsaker** [comment]
I think this just got lost in the shuffle of everything going on at once, honestly. We'll review this again Monday (in about 32 hours, give or take, from when this comment gets posted), and probably get it merged. We'll still have to evaluate the compile hook described above, before including it as well.

As to the Composer plugin you linked above, could you open a separate issue for that one, so we can properly explore that idea in full? Thanks in advance!

**Comment by ThePixelDeveloper** [comment]
Any updates on this? The current method of composer cloning all the dependencies takes a long time.

**Comment by florentdestremau** [comment]
how is this issue going ? I'm super interested in having the custom options for composer

**Comment by sanderson** [comment]
This is now available through the `composer_install` engine config option

```yaml
run.config:
  engine: php
  engine.config:
    composer_install: 'composer install --prefer-source --no-interaction'
```

Resolved by adfe8277abd9e8363f3d53b4a8ef044b3e8550ae
